Add memory arguments for FuncDefs.

cgyurgyik opened the pull request:
Addresses #348 .
Addresses #352.

cgyurgyik commented:
Current progress:
`fuse a.fuse -b=futil`

```
def add_one(a: ubit<32>[1][1]): ubit<32> = {
   let z: ubit<1> = (0 as ubit<1>);
   let _temp = (1 as ubit<32>) + a[z][z];
   return _temp;
}

decl X: ubit<32>[1][1];

let ret: ubit<32> = add_one(X);
```

=>

```
component add_one(a_read_data: 32, a_done: 1) -> (a_write_data: 32, a_write_en: 1, a_addr0: 1, a_addr1: 1, out: 32) {
  cells {
    _temp0 = std_reg(32);
    add0 = std_add(32);
    const0 = std_const(1,0);
    const1 = std_const(32,1);
    z0 = std_reg(1);
  }
  wires {
    group let1<"static"=1> {
      z0.in = const0.out;
      z0.write_en = 1'd1;
      let1[done] = z0.done;
    }
    group let2<"static"=1> {
      _temp0.in = add0.out;
      _temp0.write_en = 1'd1;
      let2[done] = _temp0.done;
      add0.left = const1.out;
      add0.right = a_read_data;
      a_addr1 = z0.out;
      a_addr0 = z0.out;
    }
    out = _temp0.out;
  }
  control {
    par {
      let1;
      let2;
    }
  }
}
component main() -> () {
  cells {
    @external(1) X = std_mem_d2(32,1,1,1,1);
    add_one0 = add_one();
    ret0 = std_reg(32);
  }
  wires {
    group let0 {
      ret0.in = add_one0.out;
      ret0.write_en = 1'd1;
      let0[done] = ret0.done;
    }
  }
  control {
    seq {
      invoke add_one0(a_read_data=X.read_data, a_done=X.done)(a_write_data=X.write_data, a_write_en=X.write_en, a_addr0=X.addr0, a_addr1=X.addr1);
      let0;
    }
  }
}
```

However, with `--lower`, I get:
```
[Type error] [Line 3, Column 34] `a' is not bound in scope.
   let _temp = (1 as ubit<32>) + a[z][z];
                                 ^

Does this program type check without the `--lower` flag? If not, please report this as an error.
```
so I need to figure out why this is the case.

rachitnigam commented:
`--pass-debug` reveals the issue is in the loop unrolling pass:
```
===============Lower unroll and bank===============
def add_one(a0_0: ubit<32>[1][1]) : ubit<32>= {
  let z_: ubit<1> = (0 as ubit<1>);
  ---
  let a_read0 = a[z][z];
  ---
  let _temp_ = ((1 as ubit<32>) + a_read0);
  ---
  return _temp_;
}
decl X0_0: ubit<32>[1][1];
let ret_: ubit<32> = add_one(X);
```

The pass does not correctly rewrite uses of `a` in the body to `a0_0`.

cgyurgyik commented:
> `--pass-debug` reveals the issue is in the loop unrolling pass:
> 
> ```
> ===============Lower unroll and bank===============
> def add_one(a0_0: ubit<32>[1][1]) : ubit<32>= {
>   let z_: ubit<1> = (0 as ubit<1>);
>   ---
>   let a_read0 = a[z][z];
>   ---
>   let _temp_ = ((1 as ubit<32>) + a_read0);
>   ---
>   return _temp_;
> }
> decl X0_0: ubit<32>[1][1];
> let ret_: ubit<32> = add_one(X);
> ```
> 
> The pass does not correctly rewrite uses of `a` in the body to `a0_0`.

Ah, debug pass would help here. Thanks. Still a small naming issue:
```
===============Hoist memory reads===============
def add_one(a: ubit<32>[1][1]) : ubit<32>let z: ubit<1> = (0 as ubit<1>);
let a_read0 = a[z][z];
let _temp = ((1 as ubit<32>) + a_read0);
return _temp;
decl X: ubit<32>[1][1];
let ret: ubit<32> = add_one(X);
================================================
[Impossible] [Enclosing(fuselang.Compiler.checkStringWithError ast):Line(61)] Pretty printer generated a program that could not be parsed after code generation.
Use the --pass-debug flag to see what the generated program looked like.. This should never trigger. Please report this as a bug.
```

Other notes for self:
1. Change `in, out` to `src, dest` for consistency with other connects.
2. Support void functions.

rachitnigam commented:
Rebase on master. It's been fixed. The problem has to do with the --lower flag incorrectly handling memory parameters in general, not just on this branch. If you open an issue, I can get around to fixing it.

cgyurgyik commented:
> Rebase on master. The parsing problem has been fixed. The problem has to do with the --lower flag incorrectly handling memory parameters in general, not just on this branch. If you open an issue, I can get around to fixing it.

Yeah I did, and yeah I was getting similar errors hence my 7 edits. Ok.

cgyurgyik commented:
Output produced without `--lower` looks sound. Waiting on #351 to produce `.expected`.

cgyurgyik commented on src/main/scala/backends/futil/FutilBackend.scala:
```diff
@@ -194,6 +185,101 @@ private class FutilBackendHelper {
+        in a memory. For example, a D1 Memory declared as (32, 1, 1)
+        would return List[("addr0", 1)].
+    */
+    def getAddrPortToWidths(typ: TArray, id: Id): List[(String, Int)] = {
```
The Pythonic approach to this is:
```
    dims = dims2id[id] # Mapping from number of dims to compId, e.g. 1 : 'std_mem_d1'
    addresses = range(0, dims)
    indices = range(dims + 1, dims << 1 + 1)
    return [(f'addr{i}', args[n]) for (i, n) in zip(addresses, indices)]
```

so we should be able to shorten this here as well.

sgpthomas reviewed (commented):
looks good generally. I agree you should rewrite `getAddrPortToWidths` to be more functional and not have the long pattern matches. Also, the `invoke-with-memories.expect` is empty and should be filled in with the actual code

cgyurgyik commented:
> looks good generally. I agree you should rewrite `getAddrPortToWidths` to be more functional and not have the long pattern matches. Also, the `invoke-with-memories.expect` is empty and should be filled in with the actual code

Fixed. I've added the output of running `fuse` without the `--lower` flag for now, but this isn't correct (given that memory names will change). Issue mentioned above.

cgyurgyik commented:
Ok I've update the test with PR #351  submitted. Quick peek says it looks good, but I'll look over it more thoroughly tomorrow, and use Relay to verify correctness.

sgpthomas commented:
LGTM! I think it's ready to merge. @rachitnigam do you want a rebase or a merge?

rachitnigam commented:
My general opinion is "Squash and merge" because it keeps commits on master green.
